workflow/controller/controller.go

```diff
@@ -625,35 +592,75 @@ func getWfPriority(obj interface{}) (int32, time.Time) {
 }
 
 func (wfc *WorkflowController) addWorkflowInformerHandler() {
-	wfc.incompleteWfInformer.AddEventHandler(
-		cache.ResourceEventHandlerFuncs{
-			AddFunc: func(obj interface{}) {
-				key, err := cache.MetaNamespaceKeyFunc(obj)
-				if err == nil {
-					wfc.wfQueue.Add(key)
-					priority, creation := getWfPriority(obj)
-					wfc.throttler.Add(key, priority, creation)
-				}
-			},
-			UpdateFunc: func(old, new interface{}) {
-				key, err := cache.MetaNamespaceKeyFunc(new)
-				if err == nil {
-					wfc.wfQueue.Add(key)
-					priority, creation := getWfPriority(new)
-					wfc.throttler.Add(key, priority, creation)
-				}
+	wfc.wfInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				wf := obj.(*unstructured.Unstructured)
+				return wf.GetLabels()[common.LabelKeyCompleted] != "true"
 			},
-			DeleteFunc: func(obj interface{}) {
-				// IndexerInformer uses a delta queue, therefore for deletes we have to use this
-				// key function.
-				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-				if err == nil {
-					wfc.wfQueue.Add(key)
-					wfc.throttler.Remove(key)
-				}
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					key, err := cache.MetaNamespaceKeyFunc(obj)
+					if err == nil {
+						wfc.wfQueue.Add(key)
+						priority, creation := getWfPriority(obj)
+						wfc.throttler.Add(key, priority, creation)
+					}
+				},
+				UpdateFunc: func(old, new interface{}) {
+					key, err := cache.MetaNamespaceKeyFunc(new)
+					if err == nil {
+						oldWf := old.(*unstructured.Unstructured)
+						newWf := new.(*unstructured.Unstructured)
+						// we can get repeated resource version which we don't need to process
+						// this is common due to how we put the workflow back
+						if oldWf.GetResourceVersion() == newWf.GetResourceVersion() {
```

Comment:
because we put back workflows on update - we want to avoid processing them twice, so we do a version check here